File: platform/sdk/conformance/terminal-v4.test.py
import importlib.util
import json
from pathlib import Path
import unittest
from hashlib import sha256

from layerx_sdk.program_wire import decode_and_verify_program_terminal, _verify_applied_legs
from layerx_sdk.verifier import AuthorizedReceiptBatch, verify_receipt_outcome

ROOT = Path(__file__).resolve().parents[3]
spec = importlib.util.spec_from_file_location('signatures', ROOT / 'platform/integrations/fastapi/layerx_fastapi/signatures.py')
module = importlib.util.module_from_spec(spec)
spec.loader.exec_module(module)


class TerminalV4(unittest.TestCase):
    def test_signed_shared_vectors(self):
        for name, expected in [('executed-v4', 'reconstructed'), ('principal-v4', 'reconstructed'),
                               ('mutated-leg-v4', None), ('executed-v3', 'recorded_terminal_root_not_locally_reconstructable')]:
            with self.subTest(name=name):
                fixture = json.loads((ROOT / f'platform/sdk/conformance/fixtures/receipt-programs-{name}.json').read_text())
                source = fixture['authorized_batch']
                authority = AuthorizedReceiptBatch(**{field: bytes.fromhex(source[field + '_hex']) for field in
                    ('batch_id', 'asset', 'previous_state_root', 'resulting_state_root', 'sequencer_public_key')})
                verified = verify_receipt_outcome(bytes.fromhex(fixture['canonical_receipt_hex']), authority, module.LayerXSignatureVerifier(), protocol_version=3)
                self.assertEqual(verified.receipt_digest.hex(), fixture['receipt_digest_hex'])
                self.assertEqual(sha256(b'LXP/v1/activity-id\0' + bytes.fromhex(fixture['signed_activity_hex'])).digest(), verified.receipt.activity_id)
                args = (bytes.fromhex(fixture['terminal_payload_hex']), bytes.fromhex(fixture['call_graph_hex']), fixture['program_id_hex'], verified.receipt.program_outcome, 3)
                if expected is None:
                    with self.assertRaisesRegex(ValueError, 'applied transfer root'):
                        decode_and_verify_program_terminal(*args)
                else:
                    self.assertEqual(decode_and_verify_program_terminal(*args).transfer_verification, expected)
                if name == 'executed-v4':
                    for length in range(len(args[0])):
                        with self.assertRaises(ValueError):
                            decode_and_verify_program_terminal(args[0][:length], *args[1:])
                    with self.assertRaises(ValueError):
                        decode_and_verify_program_terminal(args[0] + b'\0', *args[1:])

    def test_empty_legs_require_zero_root(self):
        _verify_applied_legs(b'', bytes(32))
        with self.assertRaises(ValueError):
            _verify_applied_legs(b'', b'\1' * 32)


if __name__ == '__main__':
    unittest.main()

File: platform/sdk/go/terminal_v4_test.go
package layerx

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"os"
	"strings"
	"testing"
)

func TestSignedTerminalV4Vectors(t *testing.T) {
	for _, name := range []string{"executed-v4", "principal-v4", "mutated-leg-v4", "executed-v3", "account-bound-v4"} {
		t.Run(name, func(t *testing.T) {
			path := "../conformance/fixtures/receipt-programs-" + name + ".json"
			if name == "account-bound-v4" {
				path = "../../../programs/fixtures/pay5/receipt-account-bound-v4.json"
			}
			raw, err := os.ReadFile(path)
			if err != nil {
				t.Fatal(err)
			}
			var vector struct {
				Canonical string                  `json:"canonical_receipt_hex"`
				Activity  string                  `json:"signed_activity_hex"`
				Program   string                  `json:"program_id_hex"`
				Digest    string                  `json:"receipt_digest_hex"`
				Terminal  string                  `json:"terminal_payload_hex"`
				Graph     string                  `json:"call_graph_hex"`
				Authority receiptFixtureAuthority `json:"authorized_batch"`
			}
			if err := json.Unmarshal(raw, &vector); err != nil {
				t.Fatal(err)
			}
			authority := AuthorizedBatch{
				BatchID: fixture32(t, vector.Authority.BatchIDHex), Asset: fixture32(t, vector.Authority.AssetHex),
				PreviousStateRoot: fixture32(t, vector.Authority.PreviousStateRootHex), ResultingStateRoot: fixture32(t, vector.Authority.ResultingStateRootHex),
				SequencerPublicKey: fixture32(t, vector.Authority.SequencerPublicKeyHex),
			}
			verified, err := VerifyReceiptOutcome(fixtureBytes(t, vector.Canonical), authority, 3)
			if err != nil {
				t.Fatal(err)
			}
			if verified.ReceiptDigest != fixture32(t, vector.Digest) || verified.Receipt.ActivityID != domainDigest([]byte("LXP/v1/activity-id\x00"), fixtureBytes(t, vector.Activity)) {
				t.Fatal("signed vector identity mismatch")
			}
			outcome := verified.Receipt.ProgramOutcome
			if outcome == nil {
				t.Fatal("missing outcome")
			}
			terminal, graph := fixtureBytes(t, vector.Terminal), fixtureBytes(t, vector.Graph)
			inner, err := unwrapAppliedProgramTerminal(terminal, *outcome)
			if name == "mutated-leg-v4" {
				if err == nil || !strings.Contains(err.Error(), "applied transfer root") {
					t.Fatalf("mutated leg: %v", err)
				}
				return
			}
			if err != nil {
				t.Fatal(err)
			}
			projection, err := decodeProgramTerminal(outcome.TerminalKind, outcome.ABIVersion, inner, fixture32(t, vector.Program), outcome.ResultCode)
			if err != nil {
				t.Fatal(err)
			}
			execution := ProgramExecutionDocument{
				ActivityID: hex.EncodeToString(verified.Receipt.ActivityID[:]), ProgramID: vector.Program,
				ModuleVersion: verified.Receipt.ModuleVersion, GuestABIVersion: outcome.ABIVersion,
				Receipt: vector.Canonical, ReceiptDigest: vector.Digest, TerminalPayload: vector.Terminal, CallGraph: vector.Graph,
				ResultCode: outcome.ResultCode, Outcome: projection.Outcome,
			}
			result, err := VerifyProgramReceipt(execution, authority, 3)
			if err != nil {
				t.Fatal(err)
			}
			expected := ProgramTransfersReconstructed
			if name == "executed-v3" {
				expected = ProgramTransfersRecorded
			}
			if result.TransferVerification != expected {
				t.Fatalf("status %q", result.TransferVerification)
			}
			if name == "executed-v4" || name == "account-bound-v4" {
				for length := 0; length < len(terminal); length++ {
					if _, err := verifyProgramTerminal(execution, verified.Receipt, terminal[:length], graph); err == nil {
						t.Fatalf("accepted truncation %d", length)
					}
				}
				if _, err := verifyProgramTerminal(execution, verified.Receipt, append(append([]byte{}, terminal...), 0), graph); err == nil {
					t.Fatal("accepted trailing byte")
				}
			}
		})
	}
}

func TestAppliedEmptyLegsRequireZeroRoot(t *testing.T) {
	encoded := append([]byte("LXP/programs/terminal-applied-legs/v1\x00"), 0, 0, 0, 1, 1, 0, 0, 0, 0)
	outcome := ProgramReceiptOutcome{EncodingVersion: 4, AppliedLegsDigest: sha256.Sum256(nil)}
	if _, err := unwrapAppliedProgramTerminal(encoded, outcome); err != nil {
		t.Fatal(err)
	}
	outcome.TransferRoot[0] = 1
	if _, err := unwrapAppliedProgramTerminal(encoded, outcome); err == nil {
		t.Fatal("accepted nonzero empty root")
	}
}

func TestNativeAccountAuthorizationVectors(t *testing.T) {
	raw, err := os.ReadFile("../../../programs/fixtures/pay5/account-authorization-vectors.json")
	if err != nil {
		t.Fatal(err)
	}
	var vectors []struct {
		Name    string
		Encoded string
		Root    string
		Accept  bool
	}
	if err := json.Unmarshal(raw, &vectors); err != nil {
		t.Fatal(err)
	}
	for _, v := range vectors {
		err := verifyProgramTransferAuthorization(fixtureBytes(t, v.Encoded), fixture32(t, v.Root))
		if (err == nil) != v.Accept {
			t.Fatalf("%s: accept=%v error=%v", v.Name, v.Accept, err)
		}
	}
}
